---
title: "BRCA-EU Simple Somatic Mutations"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(tidyverse)
library(fuzzyjoin)
library(knitr)
library(scales)
library(GenomicRanges)
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg19)
```

```{r functions}
isOverlap <- function(intervals, is.bed.format = F){
  gr <- GRanges(seqnames = Rle(intervals$chromosome), 
                ranges = IRanges(intervals$start + is.bed.format, intervals$end))
  
  print(ifelse(isDisjoint(gr), 
               "there is no overlap between intervals", 
               "there are overlaps between intervals"))
}

makeNonOverlap <- function(intervals, is.bed.format = F){
  gr <- GRanges(seqnames = Rle(intervals$chromosome), 
                ranges = IRanges(intervals$start + is.bed.format, intervals$end))
  gr_reduced <- GenomicRanges::reduce(gr)
  outcome <- tibble(
    chromosome = as.character(seqnames(gr_reduced)),
    start = start(gr_reduced),
    end = end(gr_reduced))
  return(outcome)
}

reverseMerge <- function(count_mat){
  features <- colnames(count_mat)
  list_drop <- NULL
  for(feature in features){
    reverse_feature <- as.character(reverseComplement(DNAString(feature)))
    if(feature < reverse_feature)
      count_mat[,feature] <- count_mat[,feature] + count_mat[,reverse_feature]
    else if(feature > reverse_feature)
      list_drop <- c(list_drop, which(features == feature))
    #if palindromic then feature == reverse_feature, and we count only once
  }
  count_mat <- count_mat[, -list_drop]
  return(count_mat)
}
```

```{r import and transform data}
# file paths and file names
icgc.data.path <- "/Users/jz132/Desktop/Gordanlab/Data/ICGC"
genomic.interval.path <- "/Users/jz132/r_projects/cancer-mutations/pelinks"
refseq.data.path <- "/Users/jz132/r_projects/cancer-mutations/pelinks/RefSeq"
filename <- "simple_somatic_mutation.open.BRCA-EU.tsv"
ignore.rna <- TRUE

genome <- getBSgenome("BSgenome.Hsapiens.UCSC.hg19")

keep_cols <- c("icgc_mutation_id", 
               "icgc_donor_id", 
               "chromosome",
               "chromosome_start",
               "chromosome_end",
               "assembly_version",
               "mutation_type",
               "reference_genome_allele",
               "mutated_from_allele",
               "mutated_to_allele",
               "consequence_type",
               "gene_affected",
               "transcript_affected",
               "sequencing_strategy")

# import ICGC data
setwd(icgc.data.path)
data_icgc_try <- read_tsv(filename, col_names = T, n_max = 5)
col_indicator <- paste(ifelse(colnames(data_icgc_try) %in% keep_cols, "?", "-"), collapse = "")
data_icgc_raw <- read_tsv(filename, col_names = T, col_types = col_indicator) %>%
  mutate(chromosome = paste0("chr", chromosome))

all_types <- data_icgc_raw %>% distinct(consequence_type) %>% pull()
nonexome <- c("intron_variant", "intragenic_variant", "upstream_gene_variant", 
              "downstream_gene_variant", "intergenic_region")
exome <- dplyr::setdiff(all_types, c(nonexome, NA))

# focus on single base substitution in WGS only
data_icgc_wgs <- data_icgc_raw %>% 
  filter(sequencing_strategy == "WGS",
         mutation_type == "single base substitution")
num_donors <- length(data_icgc_wgs %>% distinct(icgc_donor_id) %>% pull())
num_mutations <- length(data_icgc_wgs %>% distinct(icgc_mutation_id, icgc_donor_id) %>% pull())
```

There are `r num_donors` donors with WGS data in the study. The total number of single mutations is `r num_mutations`. Here, we use enhancers from the FANTOM project, exons from RefSeq, and promoters inferred from RefSeq TSS. The three types of genomic region are defined so that there is no overlap among them. We also take 100,000 random intervals from human genome, the lengths of which are 1,000 bp each. Here we compare the mutation rates across each type of region.

## Mutation rates in different genomic regions

```{r fantom annotation}
# import genomic coordinates of promoters, enhancers, and exons
setwd(genomic.interval.path)
data_enhancers_fantom <- read_delim("all_enhancers_fantom.txt", delim = "\t",
                                    col_names = c("chromosome", "start", "end", "enhancer"))

data_promoters_refseq <- read_delim("all_promoters_refseq.txt", delim = "\t", 
                                col_names = c("chromosome", "start", "end", "promoter")) %>%
  arrange(chromosome, start, end) %>%
  distinct()

# data_promoters_refseq <- data_promoters_refseq %>%
#   mutate(end = end - 1000) # for now, we only consider the part upstream of tss

data_exons_refseq <- read_delim("all_exons_refseq.txt", delim = "\t", 
                                col_names = c("chromosome", "start", "end", "exon"))

data_random_regions <- read_delim("random_genomic_intervals_l_1000_seed_1234_g_hg19.bed", 
                                  delim = "\t", 
                                  col_names = F) %>% 
  select(chromosome = 1, start = 2, end = 3)

data_eplinks_short <- read_delim("eplinks-fantom-filtered.csv", delim = ",")
data_eplinks_long <- data_eplinks_short %>%
  mutate(tss = strsplit(tss, ";")) %>%
  unnest(tss)

data_pelinks <- read_csv("pelinks-fantom-filtered.csv")

# to calculate mutation rate we need non-overlapping genomic intervals
data_enhancers_fantom_nolap <- makeNonOverlap(data_enhancers_fantom, is.bed.format = T)
data_promoters_refseq_nolap <- makeNonOverlap(data_promoters_refseq, is.bed.format = T)
data_exons_refseq_nolap <- makeNonOverlap(data_exons_refseq, is.bed.format = T)
data_random_regions_nolap <- makeNonOverlap(data_random_regions, is.bed.format = T)

data_exons_in_promoters <- data_exons_refseq_nolap %>%
  genome_inner_join(data_promoters_refseq_nolap) %>%
  mutate(start = pmax(start.x, start.y),
         end = pmin(end.x, end.y)) %>%
  select(chromosome = chromosome.x, start, end) %>%
  distinct()

length_enhancers <- sum(data_enhancers_fantom_nolap$end - data_enhancers_fantom_nolap$start)
length_exons <- sum(data_exons_refseq_nolap$end - data_exons_refseq_nolap$start)
length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
length_promoters <- sum(data_promoters_refseq_nolap$end - data_promoters_refseq_nolap$start) - length_exons_in_promoters
length_random <- sum(data_random_regions_nolap$end -data_random_regions_nolap$start)

table_length <- tibble(position = c("promoter", "enhancer", "exon", "others", "random"),
                       length = c(length_promoters, length_enhancers, length_exons, NA, length_random))

# annotate the mutations by the types of genomic intervals
data_icgc_wgs_to_join <- data_icgc_wgs %>%
  select(chromosome = chromosome,
         start = chromosome_start,
         end = chromosome_end,
         icgc_mutation_id,
         icgc_donor_id,
         ref = reference_genome_allele,
         mut = mutated_to_allele) %>%
  distinct()

# enhancer mutations
mut_enhancer <- data_enhancers_fantom_nolap %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut) %>%
  distinct()

# exon mutations
mut_exon <- data_exons_refseq_nolap %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut) %>%
  distinct()

# promoter mutations
mut_promoter <- data_promoters_refseq_nolap %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut) %>%
  distinct() %>%
  dplyr::setdiff(mut_exon)

mut_others <- data_icgc_wgs_to_join %>%
  dplyr::setdiff(bind_rows(mut_exon, mut_promoter, mut_enhancer))

mut_random <- data_random_regions_nolap %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

data_icgc_wgs_pos_fantom <- mutate(mut_exon, position = "exon") %>%
  bind_rows(mutate(mut_promoter, position = "promoter")) %>%
  bind_rows(mutate(mut_enhancer, position = "enhancer")) %>%
  bind_rows(mutate(mut_others, position = "others")) %>%
  bind_rows(mutate(mut_random, position = "random")) %>%
  select(icgc_mutation_id, icgc_donor_id, position)
```

The number of single mutations in each type of region:

```{r mutation position table, results='asis'}
table_mutation_pos_fantom <- data_icgc_wgs_pos_fantom %>% 
  group_by(position) %>%
  tally(name = "count") %>%
  inner_join(table_length) %>%
  mutate(mutation_rate = scientific(count/num_donors/length))

kable(table_mutation_pos_fantom)
```

## Mutation rates for different promoter length

Promoter regions are usually defined as the parts of genome adjacent to each TSS. The exact length to the left and right of the TSS varies from study to study. We want to investigate if the definition of promoter will affect the mutation rate estimation, so we've tried different length and computed the corresponding rate.

```{r promoter length, results='asis'}
setwd(genomic.interval.path)
data_promoters_refseq_full <- read_delim("all_promoters_refseq.txt", delim = "\t", 
                                        col_names = c("chromosome", "start", "end", "promoter", "strand"))
data_promoters_test <- list("l1000_r1000" = data_promoters_refseq_full,
                            "l1000_r0" = data_promoters_refseq_full %>% 
                              mutate(end = ifelse(strand == "+", end - 1000, end),
                                     start = ifelse(strand == "-", start + 1000, start)),
                            "l500_r500" =  data_promoters_refseq_full %>%
                              mutate(start = start + 500, end = end - 500),
                            "l5000_r1000" = data_promoters_refseq_full %>% 
                              mutate(start = ifelse(strand == "+", start - 4000, start),
                                     end = ifelse(strand == "-", end + 4000, end)),
                            "l10000_r1000" = data_promoters_refseq_full %>% 
                              mutate(start = ifelse(strand == "+", start - 9000, start),
                                     end = ifelse(strand == "-", end + 9000, end))
)

mut_count <- NULL
length_promoters_all <- NULL
for(data_promoters in data_promoters_test){
  data_promoters <- makeNonOverlap(data_promoters, is.bed.format = T)
  data_exons_in_promoters <- data_exons_refseq_nolap %>%
    genome_inner_join(data_promoters) %>%
    mutate(start = pmax(start.x, start.y), end = pmin(end.x, end.y)) %>% 
    select(chromosome = chromosome.x, start, end) %>%
    distinct()
  
  length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
  length_promoters_raw <- sum(data_promoters$end - data_promoters$start)
  length_promoters <- length_promoters_raw - length_exons_in_promoters
  
  mut_promoter <- data_promoters %>% 
    genome_left_join(data_icgc_wgs_to_join) %>%
    na.omit() %>%
    select(chromosome = chromosome.y,
           start = start.y,
           end = end.y,
           icgc_mutation_id,
           icgc_donor_id,
           ref,
           mut) %>%
    distinct() %>%
  dplyr::setdiff(mut_exon)
  
  mut_count <- c(mut_count, nrow(mut_promoter))
  length_promoters_all <- c(length_promoters_all, length_promoters)
}

table_promoters_test <- tibble(
  promoter = names(data_promoters_test),
  count = mut_count,
  length = length_promoters_all,
  mutation_rate = scientific(mut_count/num_donors/length, digits = 3)
)
kable(table_promoters_test)
```

## Exon mutations

```{r calculate exon background}
# number of mutation in each exon
data_exons_mutated <- data_exons_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  select(chromosome = chromosome.x,
         start = start.x,
         end = end.x,
         exon,
         icgc_mutation_id,
         icgc_donor_id) %>%
  mutate(count = ifelse(is.na(icgc_mutation_id), 0, 1)) %>%
  group_by(chromosome, start, end, exon) %>%
  summarise(count = sum(count)) %>%
  mutate(length = end - start + 1) %>%
  ungroup()

mut_exon <- data_exons_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut)
mut_exon_bg <- mut_exon %>%
  mutate(start = start - 1, end = end + 1)

# exon sequences and trinucleotide frequencies
seq_exons_refseq <- getSeq(genome, names = data_exons_mutated$chromosome,
                               start = data_exons_mutated$start,
                               end = data_exons_mutated$end)
mat_tri <- reverseMerge(trinucleotideFrequency(seq_exons_refseq))
freq_tri <- enframe(colSums(mat_tri), name = "trinucleotide", value = "count")

# frequencies of trinucleotides that are mutated
seq_exon_mutations_ref <- getSeq(genome, names = mut_exon_bg$chromosome,
                                     start = mut_exon_bg$start,
                                     end = mut_exon_bg$end)
mat_tri_mut <- reverseMerge(trinucleotideFrequency(seq_exon_mutations_ref))
freq_tri_mut <- enframe(colSums(mat_tri_mut), name = "trinucleotide", value = "mut_count") %>%
  arrange(desc(mut_count))

freq_tri <- freq_tri %>%
  inner_join(freq_tri_mut) %>%
  mutate(mut_rate = mut_count/count/num_donors)

data_exons_mutated <- data_exons_mutated %>%
  mutate(expected_count = as.vector(mat_tri%*%freq_tri$mut_rate*num_donors),
         var_count = as.vector(mat_tri%*%(freq_tri$mut_rate*(1-freq_tri$mut_rate))*num_donors))
```

<!-- The transcripts whose exon harbor the greatest number of mutations: -->

```{r top mutated tss exon}
# annotate by refseq 
setwd(refseq.data.path)
mapping_hgnc <- read_delim("refseq_to_hgnc.txt", delim = "\t") %>%
  select(gene = `Approved symbol`,
         refseq_id = `RefSeq IDs`,
         refseq_id_ncbi = `RefSeq(supplied by NCBI)`) %>%
  pivot_longer(cols = refseq_id:refseq_id_ncbi, values_to = "tss") %>%
  select(gene, tss) %>%
  na.omit() %>%
  distinct()
mapping_refseq_exon <- read_delim("refseq_exons_171007.bed", delim = '\t',
                             col_names = F) %>%
  select(chromosome = "X1", start = "X2", end = "X3", tss = "X4") %>%
  mutate(tss = gsub("_exon.*$", "", tss))

data_exons_mutated_annotated <- data_exons_mutated %>%
  inner_join(mapping_refseq_exon) %>%
  inner_join(mapping_hgnc)

if(ignore.rna){
  data_exons_mutated_annotated <- data_exons_mutated_annotated %>% filter(grepl("NM", tss))
}

# per tss
table_exon_mutation_by_tss <- data_exons_mutated_annotated %>%
  select(-gene) %>% 
  distinct() %>%
  group_by(tss) %>%
  summarise(count = sum(count),
            exon_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

# kable(table_exon_mutation_by_tss %>% dplyr::slice(1:20))
```

The genes whose exons harbor the greatest number of mutations:

```{r top mutated genes exon}
table_exon_mutation_by_gene <- data_exons_mutated_annotated %>%
  select(-tss) %>% 
  distinct() %>%
  group_by(gene) %>%
  summarise(count = sum(count),
            exon_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

kable(table_exon_mutation_by_gene %>% arrange(desc(log.p_value)) %>% dplyr::slice(1:20))
```

## Promoter mutations

```{r calculate promoter background}
# number of mutation in each promoter
data_promoters_mutated <- data_promoters_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  select(chromosome = chromosome.x,
         start = start.x,
         end = end.x,
         promoter,
         icgc_mutation_id,
         icgc_donor_id) %>%
  mutate(count = ifelse(is.na(icgc_mutation_id), 0, 1)) %>%
  group_by(chromosome, start, end, promoter) %>%
  summarise(count = sum(count)) %>%
  mutate(length = end - start + 1) %>%
  ungroup()

mut_promoter <- data_promoters_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut)
mut_promoter_bg <- mut_promoter %>%
  mutate(start = start - 1, end = end + 1)

# promoter sequences and trinucleotide frequencies
seq_promoters_refseq <- getSeq(genome, names = data_promoters_mutated$chromosome,
                               start = data_promoters_mutated$start,
                               end = data_promoters_mutated$end)
mat_tri <- reverseMerge(trinucleotideFrequency(seq_promoters_refseq))
freq_tri <- enframe(colSums(mat_tri), name = "trinucleotide", value = "count")

# frequencies of trinucleotides that are mutated
seq_promoter_mutations_ref <- getSeq(genome, names = mut_promoter_bg$chromosome,
                                     start = mut_promoter_bg$start,
                                     end = mut_promoter_bg$end)
mat_tri_mut <- reverseMerge(trinucleotideFrequency(seq_promoter_mutations_ref))
freq_tri_mut <- enframe(colSums(mat_tri_mut), name = "trinucleotide", value = "mut_count") %>%
  arrange(desc(mut_count))

freq_tri <- freq_tri %>%
  inner_join(freq_tri_mut) %>%
  mutate(mut_rate = mut_count/count/num_donors)

data_promoters_mutated <- data_promoters_mutated %>%
    mutate(expected_count = as.vector(mat_tri%*%freq_tri$mut_rate*num_donors),
         var_count = as.vector(mat_tri%*%(freq_tri$mut_rate*(1-freq_tri$mut_rate))*num_donors))
```

<!-- The transcripts whose promoters harbor the greatest number of mutations: -->

```{r top mutated tss promoter, echo=F}
# annotate by refseq 
setwd(refseq.data.path)
mapping_hgnc <- read_delim("refseq_to_hgnc.txt", delim = "\t") %>%
  select(gene = `Approved symbol`,
         refseq_id = `RefSeq IDs`,
         refseq_id_ncbi = `RefSeq(supplied by NCBI)`) %>%
  pivot_longer(cols = refseq_id:refseq_id_ncbi, values_to = "tss") %>%
  select(gene, tss) %>%
  na.omit() %>%
  distinct()
mapping_refseq_tss <- read_delim("refseq_TSS_hg19_170929.bed", delim = '\t',
                             col_names = F) %>%
  select(c(1,2,4)) %>%
  dplyr::rename(chromosome = "X1", pos = "X2", tss = "X4")

data_promoters_mutated_annotated <- data_promoters_mutated %>%
  mutate(pos = end - 1000) %>%
  inner_join(mapping_refseq_tss) %>%
  inner_join(mapping_hgnc) %>%
  select(-pos)

if(ignore.rna){
  data_promoters_mutated_annotated <- data_promoters_mutated_annotated %>% filter(grepl("NM", tss))
}

# per tss
table_promoter_mutation_by_tss <- data_promoters_mutated_annotated %>%
  select(-gene) %>% 
  distinct() %>%
  group_by(tss) %>%
  summarise(count = sum(count),
            promoter_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

# kable(table_promoter_mutation_by_tss %>% dplyr::slice(1:20))
```

The genes whose promoters harbor the greatest number of mutations:

```{r top mutated genes promoter, echo=F}
table_promoter_mutation_by_gene <- data_promoters_mutated_annotated %>%
  select(-tss) %>% 
  distinct() %>%
  group_by(gene) %>%
  summarise(count = sum(count),
            promoter_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

kable(table_promoter_mutation_by_gene %>% arrange(desc(log.p_value)) %>% dplyr::slice(1:20))
```

## Enhancer mutations

```{r calculate enhancer background}
# number of mutation in each enhancer
data_enhancers_mutated <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  select(chromosome = chromosome.x,
         start = start.x,
         end = end.x,
         enhancer,
         icgc_mutation_id,
         icgc_donor_id) %>%
  mutate(count = ifelse(is.na(icgc_mutation_id), 0, 1)) %>%
  group_by(chromosome, start, end, enhancer) %>%
  summarise(count = sum(count)) %>%
  mutate(length = end - start + 1) %>%
  ungroup()

mut_enhancer <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id,
         ref,
         mut)
mut_enhancer_bg <- mut_enhancer %>%
  mutate(start = start - 1, end = end + 1)

# the tri-nucleotide background of the mutations
seq_enhancers_fantom <- getSeq(genome, names = data_enhancers_fantom$chromosome,
                               start = data_enhancers_fantom$start,
                               end = data_enhancers_fantom$end)
mat_tri <- reverseMerge(trinucleotideFrequency(seq_enhancers_fantom))
freq_tri <- enframe(colSums(mat_tri), name = "trinucleotide", value = "count")

seq_enhancer_mutations_ref <- getSeq(genome, names = mut_enhancer_bg$chromosome,
                                 start = mut_enhancer_bg$start,
                                 end = mut_enhancer_bg$end)
mat_tri_mut <- reverseMerge(trinucleotideFrequency(seq_enhancer_mutations_ref))
freq_tri_mut <- enframe(colSums(mat_tri_mut), name = "trinucleotide", value = "mut_count") %>%
  arrange(desc(mut_count))
freq_tri <- freq_tri %>%
  inner_join(freq_tri_mut) %>%
  mutate(mut_rate = mut_count/count/num_donors)

data_enhancers_mutated <- data_enhancers_mutated %>%
  mutate(expected_count = as.vector(mat_tri%*%freq_tri$mut_rate*num_donors),
         var_count = as.vector(mat_tri%*%(freq_tri$mut_rate*(1-freq_tri$mut_rate))*num_donors))
```

<!-- The transcripts whose enhancers harbor the greatest number of mutations: -->

```{r top mutated tss}
# per tss
data_enhancers_mutated_annotated_tss <- data_enhancers_mutated %>%
  inner_join(data_eplinks_long)
table_enhancer_mutation_by_tss <- data_enhancers_mutated_annotated_tss %>%
  select(-gene) %>% 
  distinct() %>%
  group_by(tss) %>%
  summarise(count = sum(count),
            enhancer_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

# kable(table_enhancer_mutation_by_tss %>% dplyr::slice(1:20))
```

The genes whose enhancers harbor the greatest number of mutations:

```{r top mutated genes}
# per gene
data_enhancers_mutated_annotated_gene <- data_enhancers_mutated %>%
  inner_join(data_eplinks_short)
table_enhancer_mutation_by_gene <- data_enhancers_mutated_annotated_gene %>%
  select(-tss) %>% 
  distinct() %>%
  group_by(gene) %>%
  summarise(count = sum(count),
            enhancer_length = sum(length),
            expected_count = sum(expected_count),
            var_count = sum(var_count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  arrange(desc(count))

kable(table_enhancer_mutation_by_gene %>% arrange(desc(log.p_value)) %>% dplyr::slice(1:20))
```

## Regulatory mutations

<!-- The transcripts whose regulatory regions have the most significant excess of mutations: -->

```{r top mutated tss all regulatory regions}
# here we use left join but please note that full join will give us 
# a different result since fantom enhancers are linked with tss
# that are no longer valid in the current Refseq
table_regulatory_mutation_by_tss <- table_promoter_mutation_by_tss %>%
  left_join(table_enhancer_mutation_by_tss, by = "tss") %>%
  mutate(count = replace_na(count.x, 0) + replace_na(count.y, 0), 
         total_length = replace_na(promoter_length, 0) + replace_na(enhancer_length, 0), 
         expected_count = replace_na(expected_count.x, 0) + replace_na(expected_count.y, 0),
         var_count = replace_na(var_count.x, 0) + replace_na(var_count.y, 0)) %>%
  arrange(desc(count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  select(tss, count, promoter_length, enhancer_length, total_length, 
         expected_count, log.p_value)

# kable(table_regulatory_mutation_by_tss %>% arrange(desc(log.p_value)) %>% dplyr::slice(1:20))
```

The genes whose regulatory regions have the most significant excess of mutations:

```{r top mutated genes all regulatory regions}
# here we use left join but please note that full join will give us 
# a different result since fantom enhancers are linked with tss
# that are no longer valid in the current Refseq
table_regulatory_mutation_by_gene <- table_promoter_mutation_by_gene %>%
  left_join(table_enhancer_mutation_by_gene, by = "gene") %>%
  mutate(count = replace_na(count.x, 0) + replace_na(count.y, 0), 
         total_length = replace_na(promoter_length, 0) + replace_na(enhancer_length, 0), 
         expected_count = replace_na(expected_count.x, 0) + replace_na(expected_count.y, 0),
         var_count = replace_na(var_count.x, 0) + replace_na(var_count.y, 0)) %>%
  arrange(desc(count)) %>%
  mutate(log.p_value = -log(dpois(count, expected_count) +
           ppois(count, expected_count, lower.tail = F), 10)) %>%
  select(gene, count, promoter_length, enhancer_length, total_length, 
         expected_count, log.p_value)

kable(table_regulatory_mutation_by_gene %>% arrange(desc(log.p_value)) %>% dplyr::slice(1:20))
```



