---
title: "LIRI-JP Simple Somatic Mutations"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(tidyverse)
library(fuzzyjoin)
library(knitr)
library(scales)
library(GenomicRanges)
```

```{r functions, echo=FALSE}
isOverlap <- function(intervals, is.bed.format = F){
  gr <- GRanges(seqnames = Rle(intervals$chromosome), 
                ranges = IRanges(intervals$start + is.bed.format, intervals$end))
  
  print(ifelse(isDisjoint(gr), 
               "there is no overlap between intervals", 
               "there are overlaps between intervals"))
}

makeNonOverlap <- function(intervals, is.bed.format = F){
  gr <- GRanges(seqnames = Rle(intervals$chromosome), 
                ranges = IRanges(intervals$start + is.bed.format, intervals$end))
  gr_reduced <- GenomicRanges::reduce(gr)
  outcome <- tibble(
    chromosome = as.character(seqnames(gr_reduced)),
    start = start(gr_reduced),
    end = end(gr_reduced))
  return(outcome)
}
```

```{r import and transform data, echo=FALSE}
# file paths and file names
icgc.data.path <- "/Users/jz132/Desktop/Gordanlab/Data/ICGC"
genomic.interval.path <- "/Users/jz132/r_projects/cancer-mutations/pelinks"
filename <- "simple_somatic_mutation.open.LIRI-JP.tsv"

keep_cols <- c("icgc_mutation_id", 
               "icgc_donor_id", 
               "chromosome",
               "chromosome_start",
               "chromosome_end",
               "assembly_version",
               "mutation_type",
               "reference_genome_allele",
               "mutated_from_allele",
               "mutated_to_allele",
               "consequence_type",
               "gene_affected",
               "transcript_affected",
               "sequencing_strategy")

# import ICGC data
setwd(icgc.data.path)
data_icgc_try <- read_tsv(filename, col_names = T, n_max = 5)
col_indicator <- paste(ifelse(colnames(data_icgc_try) %in% keep_cols, "?", "-"), collapse = "")
data_icgc_raw <- read_tsv(filename, col_names = T, col_types = col_indicator) %>%
  mutate(chromosome = paste0("chr", chromosome))

all_types <- data_icgc_raw %>% distinct(consequence_type) %>% pull()
nonexome <- c("intron_variant", "intragenic_variant", "upstream_gene_variant", 
              "downstream_gene_variant", "intergenic_region")
exome <- dplyr::setdiff(all_types, c(nonexome, NA))

# focus on single base substitution in WGS only
data_icgc_wgs <- data_icgc_raw %>% 
  filter(sequencing_strategy == "WGS", mutation_type == "single base substitution")
num_donors <- length(data_icgc_wgs %>% distinct(icgc_donor_id) %>% pull())
num_mutations <- length(data_icgc_wgs %>% distinct(icgc_mutation_id, icgc_donor_id) %>% pull())
```

There are `r num_donors` donors with WGS data in the study. The total number of single mutations is `r num_mutations`. Here, we use enhancers from the FANTOM project, exons from RefSeq, and promoters inferred from RefSeq tss. The three types of genomic regions are defined so that there is no overlap among them. We also take 100,000 random intervals from human genome. Each interval is 1,000 bp long.

```{r fantom annotation, echo=FALSE}
# import genomic coordinates of promoters, enhancers, and exons
setwd(genomic.interval.path)
data_enhancers_fantom <- read_delim("all_enhancers_fantom.txt", delim = "\t",
                                    col_names = c("chromosome", "start", "end", "enhancer"))
data_promoters_refseq <- read_delim("all_promoters_refseq.txt", delim = "\t",
                                    col_names = c("chromosome", "start", "end", "promoter"))
data_exons_refseq <- read_delim("all_exons_refseq.txt", delim = "\t", 
                                col_names = c("chromosome", "start", "end", "exon"))
data_random_regions <- read_delim("random_genomic_intervals_l_1000_seed_1234_g_hg19.bed", 
                                  delim = "\t", 
                                  col_names = F) %>% select(1:3)
colnames(data_random_regions) <- c("chromosome", "start", "end")

# to calculate mutation rate we need non-overlap genomic intervals
data_promoters_refseq <- makeNonOverlap(data_promoters_refseq)
data_exons_refseq <- makeNonOverlap(data_exons_refseq)
data_random_regions <- makeNonOverlap(data_random_regions, is.bed.format = T)

data_exons_in_promoters <- data_exons_refseq %>%
  genome_inner_join(data_promoters_refseq) %>%
  mutate(start = pmax(start.x, start.y),
         end = pmin(end.x, end.y)) %>%
  select(chromosome = chromosome.x, start, end) %>%
  distinct()

length_enhancers <- sum(data_enhancers_fantom$end - data_enhancers_fantom$start)
length_exons <- sum(data_exons_refseq$end - data_exons_refseq$start)
length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
length_promoters <- sum(data_promoters_refseq$end - data_promoters_refseq$start) - length_exons_in_promoters
length_random <- sum(data_random_regions$end -data_random_regions$start)

table_length <- tibble(position = c("promoter", "enhancer", "exon", "others", "random"),
                       length = c(length_promoters, length_enhancers, length_exons, NA, length_random))

# annotate the mutations by the types of genomic intervals
data_icgc_wgs_to_join <- data_icgc_wgs %>%
  filter(!is.na(consequence_type)) %>% 
  select(chromosome = chromosome,
         start = chromosome_start,
         end = chromosome_end,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_enhancer <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_exon <- data_exons_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_promoter <- data_promoters_refseq %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct() %>%
  dplyr::setdiff(mut_exon)

mut_others <- data_icgc_wgs_to_join %>%
  dplyr::setdiff(bind_rows(mut_exon, mut_promoter, mut_enhancer))

mut_random <- data_random_regions %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

data_icgc_wgs_pos_fantom <- mutate(mut_exon, position = 'exon') %>%
  bind_rows(mutate(mut_promoter, position = 'promoter')) %>%
  bind_rows(mutate(mut_enhancer, position = 'enhancer')) %>%
  bind_rows(mutate(mut_others, position = 'others')) %>%
  bind_rows(mutate(mut_random, position = 'random')) %>%
  select(icgc_mutation_id, icgc_donor_id, position)
```

The number of single mutations in each type of region:

```{r mutation pos table, echo=FALSE, results='asis'}
table_mutation_pos_fantom <- data_icgc_wgs_pos_fantom %>% 
  group_by(position) %>%
  tally(name = "count") %>%
  inner_join(table_length) %>%
  mutate(mutation_rate = scientific(count/num_donors/length))
kable(table_mutation_pos_fantom)
```

For promoter regions, we played with its length:

```{r promoter length, echo=F, results='asis'}
setwd(genomic.interval.path)
data_promoters_refseq <- read_delim("all_promoters_refseq.txt", delim = "\t",
                                    col_names = c("chromosome", "start", "end"))
data_promoters_test <- list("l1000_r1000" = data_promoters_refseq,
                            "l1000_r0" = data_promoters_refseq %>% 
                              mutate(end = end - 1000),
                            "l500_r500" =  data_promoters_refseq %>%
                              mutate(start = start + 500, end = end - 500),
                            "l5000_r1000" = data_promoters_refseq %>% 
                              mutate(start = start - 4000),
                            "l10000_r1000" = data_promoters_refseq %>%
                              mutate(start = start - 9000))

mut_count <- NULL
length_promoters_all <- NULL
for(data_promoters in data_promoters_test){
  data_promoters <- makeNonOverlap(data_promoters)
  data_exons_in_promoters <- data_exons_refseq %>%
    genome_inner_join(data_promoters) %>%
    mutate(start = pmax(start.x, start.y), end = pmin(end.x, end.y)) %>% 
    select(chromosome = chromosome.x, start, end) %>%
    distinct()
  
  length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
  length_promoters_raw <- sum(data_promoters$end - data_promoters$start)
  length_promoters <- length_promoters_raw - length_exons_in_promoters
  
  mut_promoter <- data_promoters %>% 
    genome_left_join(data_icgc_wgs_to_join) %>%
    na.omit() %>%
    select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
    distinct() %>%
    dplyr::setdiff(mut_exon)
  
  mut_count <- c(mut_count, nrow(mut_promoter))
  length_promoters_all <- c(length_promoters_all, length_promoters)
}

table_promoters_test <- tibble(
  promoter = names(data_promoters_test),
  count = mut_count,
  length = length_promoters_all,
  mutation_rate = scientific(mut_count/num_donors/length, digits = 3)
)
kable(table_promoters_test)
```

The top-mutated enhancers:

```{r top mutated enhancers, echo=F}
data_enhancers_mutated <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  select(chromosome = chromosome.x,
         start = start.x,
         end = end.x,
         enhancer,
         icgc_mutation_id,
         icgc_donor_id) %>%
  mutate(count = ifelse(is.na(icgc_mutation_id), 0, 1)) %>%
  group_by(chromosome, start, end, enhancer) %>%
  summarise(count = sum(count)) %>%
  mutate(length = end - start) %>%
  mutate(mut_rate = count/length) %>%
  ungroup() %>%
  arrange(desc(count))

kable(data_enhancers_mutated %>% dplyr::slice(1:10))
```

The tss whose enhancers harbor the greatest number of mutations:

```{r top mutated tss, echo=F}
setwd(genomic.interval.path)
data_eplinks <- read_delim("eplinks-fantom-filtered.csv", delim = ",") %>%
  mutate(tss = strsplit(tss, ";")) %>%
  unnest(tss)
data_enhancers_per_tss <- data_eplinks %>% 
  group_by(tss) %>%
  tally(name = "e_count") %>%
  ungroup()

data_tss_mutated <- data_enhancers_mutated %>%
  inner_join(data_eplinks) %>%
  group_by(tss) %>%
  summarise(mut_count = sum(count)) %>%
  ungroup() %>%
  arrange(desc(mut_count)) %>%
  inner_join(data_enhancers_per_tss)

kable(data_tss_mutated %>% dplyr::slice(1:10))
```

The genes whose enhancers harbor the greatest number of mutations:

```{r top mutated genes, echo=F}
setwd(genomic.interval.path)
data_eplinks <- read_delim("eplinks-fantom-filtered.csv", delim = ",")
data_enhancers_per_gene <- data_eplinks %>% 
  group_by(gene) %>%
  tally(name = "e_count") %>%
  ungroup()

data_genes_mutated <- data_enhancers_mutated %>%
  inner_join(data_eplinks) %>%
  group_by(gene) %>%
  summarise(mut_count = sum(count)) %>%
  ungroup() %>%
  arrange(desc(mut_count)) %>%
  inner_join(data_enhancers_per_gene)

kable(data_genes_mutated %>% dplyr::slice(1:10))
```
