---
title: "LIRI-JP Simple Somatic Mutations"
output: github_document
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(tidyverse)
library(fuzzyjoin)
library(knitr)
library(scales)
library(GenomicRanges)
```

```{r functions, echo=FALSE}
nonOverlap <- function(intervals){
  gr <- GRanges(seqnames = Rle(intervals$chromosome), 
                ranges = IRanges(intervals$start, intervals$end))
  gr_reduced <- GenomicRanges::reduce(gr)
  outcome <- tibble(
    chromosome = as.character(seqnames(gr_reduced)),
    start = start(gr_reduced),
    end = end(gr_reduced))
  return(outcome)
}
```

```{r import and transform data, echo=FALSE}
# file paths and file names
icgc.data.path <- "/Users/jz132/Desktop/Gordanlab/Data/ICGC"
genomic.interval.path <- "/Users/jz132/r_projects/cancer-mutations/pelinks"
filename <- "simple_somatic_mutation.open.LIRI-JP.tsv"

keep_cols <- c("icgc_mutation_id", 
               "icgc_donor_id", 
               "chromosome",
               "chromosome_start",
               "chromosome_end",
               "assembly_version",
               "mutation_type",
               "reference_genome_allele",
               "mutated_from_allele",
               "mutated_to_allele",
               "consequence_type",
               "gene_affected",
               "transcript_affected",
               "sequencing_strategy")

# import ICGC data
setwd(icgc.data.path)
data_icgc_try <- read_tsv(filename, col_names = T, n_max = 5)
col_indicator <- paste(ifelse(colnames(data_icgc_try) %in% keep_cols, "?", "-"), collapse = "")
data_icgc_raw <- read_tsv(filename, col_names = T, col_types = col_indicator) %>%
  mutate(chromosome = paste0("chr", chromosome))

all_types <- data_icgc_raw %>% distinct(consequence_type) %>% pull()
nonexome <- c("intron_variant", "intragenic_variant", "upstream_gene_variant", 
              "downstream_gene_variant", "intergenic_region")
exome <- dplyr::setdiff(all_types, c(nonexome, NA))

# focus on single base substitution in WGS only
data_icgc_wgs <- data_icgc_raw %>% 
  filter(sequencing_strategy == "WGS", mutation_type == "single base substitution")
num_donors <- length(data_icgc_wgs %>% distinct(icgc_donor_id) %>% pull())
num_mutations <- length(data_icgc_wgs %>% distinct(icgc_mutation_id, icgc_donor_id) %>% pull())
```

There are `r num_donors` donors with WGS data in the study. The total number of single mutations is `r num_mutations`. 

```{r fantom annotation, echo=FALSE}
# import genomic coordinates of promoters, enhancers, and exons
setwd(genomic.interval.path)
data_enhancers_fantom <- read_delim("all_enhancers_fantom.bed", delim = "\t",
                                    col_names = c("chromosome", "start", "end"))
data_promoters_fantom <- read_delim("all_promoters_fantom.bed", delim = "\t",
                                    col_names = c("chromosome", "start", "end"))
data_exons_fantom <- read_delim("all_exons_fantom.bed", delim = "\t", 
                                col_names = c("chromosome", "start", "end"))
data_promoters_fantom <- nonOverlap(data_promoters_fantom)
data_exons_fantom <- nonOverlap(data_exons_fantom)
data_exons_in_promoters <- data_exons_fantom %>%
  genome_inner_join(data_promoters_fantom) %>%
  mutate(start = pmax(start.x, start.y),
         end = pmin(end.x, end.y)) %>%
  select(chromosome = chromosome.x, start, end) %>%
  distinct()
data_random_regions <- read_delim("random_genomic_intervals_l_1000_seed_1234_g_hg19.bed", 
                                  delim = "\t", 
                                  col_names = c("chromosome", "start", "end", NULL, NULL, NULL))
data_random_regions <- nonOverlap(data_random_regions)

length_enhancers <- sum(data_enhancers_fantom$end - data_enhancers_fantom$start)
length_exons <- sum(data_exons_fantom$end - data_exons_fantom$start)
length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
length_promoters <- sum(data_promoters_fantom$end - data_promoters_fantom$start) - length_exons_in_promoters
length_random <- sum(data_random_regions$end -data_random_regions$start)

table_length <- tibble(position = c("promoter", "enhancer", "exon", "others", "random"),
                       length = c(length_promoters, length_enhancers, length_exons, NA, length_random))

# check how fantom annotates the mutations
data_icgc_wgs_to_join <- data_icgc_wgs %>%
  filter(!is.na(consequence_type)) %>% 
  select(chromosome = chromosome,
         start = chromosome_start,
         end = chromosome_end,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_enhancer <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_exon <- data_exons_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

mut_promoter <- data_promoters_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct() %>%
  dplyr::setdiff(mut_exon)

mut_others <- data_icgc_wgs_to_join %>%
  dplyr::setdiff(bind_rows(mut_exon, mut_promoter, mut_enhancer))

mut_random <- data_random_regions %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  na.omit() %>%
  select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
  distinct()

data_icgc_wgs_pos_fantom <- mutate(mut_exon, position = 'exon') %>%
  bind_rows(mutate(mut_promoter, position = 'promoter')) %>%
  bind_rows(mutate(mut_enhancer, position = 'enhancer')) %>%
  bind_rows(mutate(mut_others, position = 'others')) %>%
  bind_rows(mutate(mut_random, position = 'random')) %>%
  select(icgc_mutation_id, icgc_donor_id, position)
```

The number of single mutations in each type of region:

```{r mutation pos table, echo=FALSE, results='asis'}
table_mutation_pos_fantom <- data_icgc_wgs_pos_fantom %>% 
  group_by(position) %>%
  tally(name = "count") %>%
  inner_join(table_length) %>%
  mutate(mutation_rate = scientific(count/num_donors/length))
kable(table_mutation_pos_fantom)
```

For promoter regions, we played with its length:

```{r promoter length, echo=F, results='asis'}
setwd(genomic.interval.path)
data_promoters_fantom <- read_delim("all_promoters_fantom.bed", delim = "\t",
                                    col_names = c("chromosome", "start", "end"))
data_promoters_test <- list("l1000_r1000" = data_promoters_fantom,
                            "l1000_r0" = data_promoters_fantom %>% 
                              mutate(end = end - 1000),
                            "l500_r500" =  data_promoters_fantom %>%
                              mutate(start = start + 500, end = end - 500),
                            "l5000_r1000" = data_promoters_fantom %>% 
                              mutate(start = start - 4000),
                            "l10000_r1000" = data_promoters_fantom %>%
                              mutate(start = start - 9000))

mut_count <- NULL
length_promoters_all <- NULL
for(data_promoters in data_promoters_test){
  data_promoters <- nonOverlap(data_promoters)
  data_exons_in_promoters <- data_exons_fantom %>%
    genome_inner_join(data_promoters) %>%
    mutate(start = pmax(start.x, start.y), end = pmin(end.x, end.y)) %>% 
    select(chromosome = chromosome.x, start, end) %>%
    distinct()
  
  length_exons_in_promoters <- sum(data_exons_in_promoters$end - data_exons_in_promoters$start)
  length_promoters_raw <- sum(data_promoters$end - data_promoters$start)
  length_promoters <- length_promoters_raw - length_exons_in_promoters
  
  mut_promoter <- data_promoters %>% 
    genome_left_join(data_icgc_wgs_to_join) %>%
    na.omit() %>%
    select(chromosome = chromosome.y,
         start = start.y,
         end = end.y,
         icgc_mutation_id,
         icgc_donor_id) %>%
    distinct() %>%
    dplyr::setdiff(mut_exon)
  
  mut_count <- c(mut_count, nrow(mut_promoter))
  length_promoters_all <- c(length_promoters_all, length_promoters)
}

table_promoters_test <- tibble(
  promoter = names(data_promoters_test),
  count = mut_count,
  length = length_promoters_all,
  mutation_rate = scientific(mut_count/num_donors/length, digits = 3)
)
kable(table_promoters_test)
```

```{r mutated enhancers, echo=F}
data_enhancers_mutated <- data_enhancers_fantom %>% 
  genome_left_join(data_icgc_wgs_to_join) %>%
  select(chromosome = chromosome.x,
         start = start.x,
         end = end.x,
         icgc_mutation_id,
         icgc_donor_id) %>%
  mutate(count = ifelse(is.na(icgc_mutation_id), 0, 1)) %>%
  group_by(chromosome, start, end) %>%
  summarise(count = sum(count)) %>%
  mutate(length = end - start) %>%
  mutate(mut_rate = count/length) %>%
  ungroup()
```

```{r top genes, echo=F}
setwd(genomic.interval.path)
data_eplinks <- read_delim("eplinks-filtered.csv", delim = ",")
data_enh_per_gene <- data_eplinks %>% 
  group_by(gene) %>%
  tally(name = "enh_count") %>%
  ungroup()

data_enhancers_affected <- data_enhancers_mutated %>%
  mutate(temp = paste(start, end, sep = "-")) %>%
  mutate(enhancer = paste(chromosome, temp, sep = ":")) %>%
  select(enhancer, count, length, mut_rate) %>%
  arrange(desc(count))

data_genes_affected <- data_enhancers_affected %>%
  inner_join(data_eplinks) %>%
  group_by(gene) %>%
  summarise(mut_count = sum(count)) %>%
  ungroup() %>%
  arrange(desc(mut_count)) %>%
  inner_join(data_enh_per_gene)
```

The top-mutated enhancers:

```{r top enhancers}
kable(data_enhancers_affected %>% dplyr::slice(1:10))
```

The genes whose enhancers harbor the greatest number of mutations:

```{r top gene}
kable(data_genes_affected %>% dplyr::slice(1:10))
```

